bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -932,6 +928,8 @@ const char *PreProcessor::getPreProcessResultBuffer(uint16_t clientId,
   const auto bufferOffset =
       (clientId - numOfReplicas_ - numOfInternalClients_) * clientMaxBatchSize_ + reqOffsetInBatch;
   LOG_TRACE(logger(), KVLOG(clientId, reqSeqNum, reqOffsetInBatch, bufferOffset));
+  if (preProcessResultBuffers_.find(bufferOffset) == preProcessResultBuffers_.end())
+    preProcessResultBuffers_[bufferOffset] = Sliver(new char[maxPreExecResultSize_], maxPreExecResultSize_);
   return preProcessResultBuffers_[bufferOffset].data();
 }
 
```

Comment:
Don't you need to do the same checks and allocations in getOngoingReqIdForClient?
In any case, you need to go through the code and check where we assume that the buffer exists. There are many places with direct access to ongoingRequests_ by the index, like: 
const auto &reqEntry = ongoingRequests_[getOngoingReqIndex(clientId, msgOffsetInBatch)];

Reply:
I don't think that ongoingRequests_ is related, the only way to get the memory is by using getPreProcessResultBuffer.
ongoingRequests_ is still initialized the same in the constructor.

Reply:
You are right, I mixed them up.